File: .leetcode/42.接雨水.c
/*
 * @lc app=leetcode.cn id=42 lang=c
 *
 * [42] 接雨水
 */

// @lc code=start
int min(int x,int y){
    return (x<y)?x:y;
}

int trap(int* height, int heightSize) {
    //代码有点长，简单写下思路
    //每个坐标值那一列接的水，就是min(left_max,right_max)减去该坐标值
    //当然前提是min要比该坐标值大

    int num=0;   //总共接num的水
    
    //left_max直接在循环里更新就行
    int left_max=0;
    //right_max，从又往左遍历，把每个坐标值的right_max都存数组里
    int right_max[heightSize];
    right_max[heightSize-1]=height[heightSize-1];
    for(int i=heightSize-2;i>=0;i--){
        if(height[i]>right_max[i+1]){
            right_max[i]=height[i];
        }else{
            right_max[i]=right_max[i+1];
        }
    }
    //计算num
    for(int i=1;i<heightSize-1;i++){
        if(left_max<height[i-1])left_max=height[i-1];
        if(min(left_max,right_max[i+1])>height[i]){
            num+=min(left_max,right_max[i+1])-height[i];
        }
    }
    return num;
}
// @lc code=end

File: .leetcode/42.接雨水.py
#
# @lc app=leetcode.cn id=42 lang=python
#
# [42] 接雨水
#

# @lc code=start
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        n=len(height)
        sum=0  #总共的雨水
        left=[0]*n   #某一列左边最高的（包括这一列）
        left[0]=height[0]

        right=[0]*n
        right[n-1]=height[n-1]#某一列右边最高的（包括这一列）
        
        max=0
        if n<=2:
            return 0
        for i in range(1,n):      #这个循环把left数组算完
            if(height[i]>left[i-1]):
                left[i]=height[i]
            else:
                left[i]=left[i-1]
         
        for i in range(n-2,-1,-1):     #这个循环把right数组算完
            if(height[i]>right[i+1]):
                right[i]=height[i]
            else:
                right[i]=right[i+1]
        
        for i in range(1,n-1):    ##这个循环算每一列能接的雨水
            if(min(left[i-1],right[i+1]) > height[i]):
                sum+=min(left[i-1],right[i+1]) - height[i]
        
        return sum

        
# @lc code=end

